app/main.py

```diff
@@ -1,6 +1,26 @@
 from typing import Callable
+from functools import wraps
 
+def cache(func: Callable):
+    """Decorator to cache results of function calls with immutable arguments."""
+    cache_storage = {}
 
-def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    @wraps(func)
+    def wrapper(*args):
+        if args in cache_storage:
+            print("Getting from cache")
+            return cache_storage[args]
+        print("Calculating new result")
+        result = func(*args)
+        cache_storage[args] = result
+        return result
+
+    return wrapper  # Return the wrapper function, not a function call!
+
+@cache
+def long_time_func(a: int, b: int, c: int) -> int:
+    return (a ** b ** c) % (a * c)
+
+@cache
+def long_time_func_2(n_tuple: tuple, power: int) -> list:
```

Comment:
The function `long_time_func_2` uses a tuple as an argument, which is immutable, so it is safe to use as a key in the cache. However, ensure that all elements within the tuple are also immutable to avoid issues.